heaps/min_heap.py

```diff
@@ -17,62 +17,97 @@ class MinHeap:
     def __init__(self):
         self.store = []
 
-
-    def add(self, key, value = None):
+    def add(self, key, value=None):
         """ This method adds a HeapNode instance to the heap
             If value == None the new node's value should be set to key
-            Time Complexity: ?
-            Space Complexity: ?
+            Time Complexity: O(logn)
+            Space Complexity: O(1)
         """
-        pass
+        if value is None:
+            value = key
+        # creating a heap node instance and appending to store
+        self.store.append(HeapNode(key, value))
+        # moving pointer
+        self.heap_up(len(self.store) - 1)
 
     def remove(self):
         """ This method removes and returns an element from the heap
             maintaining the heap structure
-            Time Complexity: ?
-            Space Complexity: ?
+            Time Complexity: O(logn)
+            Space Complexity: O(1)
         """
-        pass
+        # if heap does not contain any nodes return None
+        if self.empty():
+            return None
 
+        self.swap(0, len(self.store) - 1)
+        # last value removed
+        popped = self.store.pop(len(self.store) - 1)
+        self.heap_down(0)
+        return popped.value
 
-    
     def __str__(self):
         """ This method lets you print the heap, when you're testing your app.
         """
         if len(self.store) == 0:
             return "[]"
         return f"[{', '.join([str(element) for element in self.store])}]"
 
-
     def empty(self):
         """ This method returns true if the heap is empty
-            Time complexity: ?
-            Space complexity: ?
+            Time complexity: O(1)
+            Space complexity: O(1)
         """
-        pass
-
+        return len(self.store) == 0
 
     def heap_up(self, index):
         """ This helper method takes an index and
-            moves the corresponding element up the heap, if 
-            it is less than it's parent node until the Heap
-            property is reestablished.
-            
-            This could be **very** helpful for the add method.
-            Time complexity: ?
-            Space complexity: ?
+            moves it up the heap, if it is less than it's parent node.
+            It could be **very** helpful for the add method.
+            Time complexity: O(n)
+            Space complexity: O(1)
+           
```

Comment:
👍  However the time complexity is O(log n) because you're not going through the entire list and space complexity O(log n) because of the call stack.